app/src/main/java/com/github/se/cyrcle/model/address/AddressRepositoryNominatim.kt

```diff
@@ -37,24 +41,39 @@ class AddressRepositoryNominatim @Inject constructor(private val client: OkHttpC
 
                   val body = response.body?.string()
                   if (body != null) {
-                    val address = addressFromBody(body)
-                    onSuccess(address)
+                    onSuccess(addressFromBody(body))
                   } else {
-                    onSuccess(Address())
+                    onSuccess(listOf(Address()))
                   }
                 }
               }
             })
   }
+  /**
+   * Extracts an address list from the response body.
+   *
+   * @param body The response body.
+   * @return The list of addresses.
+   */
+  private fun addressFromBody(body: String): List<Address> {
 
-  /** Extracts an address from the response body. */
-  private fun addressFromBody(body: String): Address {
     val jsonArray = JSONArray(body)
     if (jsonArray.length() == 0) {
-      return Address()
+      return listOf(Address())
     }
-    val jsonObject = jsonArray.getJSONObject(0)
-    val address = jsonObject.getJSONObject("address")
-    return Gson().fromJson(address.toString(), Address::class.java)
+    val addressList = mutableListOf<Address>()
+    for (i in 0 until jsonArray.length()) {
+      if (jsonArray.getJSONObject(i) != null) {
+        val jsonObject = jsonArray.getJSONObject(i)
+        val address = jsonObject.getJSONObject("address")
```

Comment:
This could have also been handled more cleanly with a custom adapter for the function, in the way it's been done in some of the ViewModels. Non-blocking but this is worth looking into...

Reply:
Yes but I took the already existing function and just added a for loop. Moreover it is coded this way inside the bootcamp. However, I can open an issue and make this a PR to rework the serialization and deserialization if you want ?

Reply:
You suggests I should create an adapter or modify an already existing one ?

Reply:
There's good references inside of our ViewModels if you want to have a look!

Reply:
Now that I tried putting an Adapter i remember why I couldn't. The `Address` data class was based on the fields that constitutes the `address` part of nominatim answer you can see. I did this way because i needed to extract the `lat` and ` lon` fields that are not part of the `address` field but rather parallel field. Does it seem clear to you ? If you have a solution to it I would gladly implement it but remember also that if the solution implies that we need to rework all the serialization/ deserialization it should be a part of a PR since this amount to a lot of work :+1:

Reply:
I agree with you on this. Let's make an issue and re-discuss this at another time. It's outside the scope of your PR as is